core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java

```diff
@@ -429,35 +432,62 @@ private static MethodDefinition generateHashBlockVectorized(ClassDefinition defi
+                "hash",
+                long.class,
+                block,
+                position);
+
+        BytecodeExpression setHashExpression;
         if (field.index() == 0) {
             // hashes[index] = hash;
-            loopBody.append(hashes.setElement(index, hash));
+            setHashExpression = hashes.setElement(index, hash);
         }
         else {
             // hashes[index] = CombineHashFunction.getHash(hashes[index], hash);
-            loopBody.append(hashes.setElement(index, invokeStatic(CombineHashFunction.class, "getHash", long.class, hashes.getElement(index), hash)));
+            setHashExpression = hashes.setElement(index, invokeStatic(CombineHashFunction.class, "getHash", long.class, hashes.getElement(index), hash));
+        }
+
+        BytecodeBlock rleHandling = new BytecodeBlock()
+                .append(new IfStatement("hash = block.isNull(position) ? NULL_HASH_CODE : hash(block, position)")
+                        .condition(block.invoke("isNull", boolean.class, position))
+                        .ifTrue(hash.set(constantLong(NULL_HASH_CODE)))
+                        .ifFalse(hash.set(computeHashNonNull)));
+        if (field.index() == 0) {
+            // Arrays.fill(hashes, 0, length, hash)
+            rleHandling.append(invokeStatic(Arrays.class, "fill", void.class, hashes, constantInt(0), length, hash));
+        }
+        else {
+            rleHandling.append(new ForLoop("for (int index = 0; index < length; index++) { hashes[index] = CombineHashFunction.getHash(hashes[index], hash); }")
+                    .initialize(index.set(constantInt(0)))
+                    .condition(lessThan(index, length))
+                    .update(index.increment())
+                    .body(setHashExpression));
         }
-        loopBody.append(position.increment());
 
-        body.append(new ForLoop("for (index = 0; index < length; index++)")
-                .initialize(index.set(constantInt(0)))
-                .condition(lessThan(index, length))
-                .update(index.increment())
-                .body(loopBody))
+        BytecodeBlock computeHashLoop = new BytecodeBlock()
+                .append(mayHaveNull.set(block.invoke("mayHaveNull", boolean.class)))
+                .append(new ForLoop("for (int index = 0; index < length; index++)")
+                        .initialize(index.set(constantInt(0)))
+                        .condition(lessThan(index, length))
+                        .update(index.increment())
+                        .body(new BytecodeBlock()
+                                .append(new IfStatement("if (mayHaveNull && block.isNull(position))")
+                                        .condition(and(mayHaveNull, block.invoke("isNull", boolean.class, position)))
+                                        .ifTrue(hash.set(constantLong(NULL_HASH_CODE)))
+                                        .ifFalse(hash.set(computeHashNonNull)))
```

Comment:
The change for `mayHaveNull` seems unrelated.  Also, you generally want to hoist that out of the loop and have two loops.

Reply:
The change is somewhat related. This is logically the same pattern that was present before special casing `RunLengthEncodedBlock`, we just assigned the value at the beginning of the method but were still using `mayHaveNull && block.isNull(position)` inside of the loop. So effectively, we've moved the assignment of `mayHaveNull` into the else-branch so that we don't need to check that for `RunLengthEncodedBlock`.

In either case, I chose to avoid splitting into two loops here to keep the bytecode size of the method small. Before C2 kicks in, we still benefit from branch prediction guessing the control flow reliably since the value of `mayHaveNull` does not change during the loop execution. After C2 kicks in, we're relying on the JIT to perform "loop unswitching" which effectively splits the single loop into two loops after removing the loop-invariant `mayHaveNull` check. You could do this explicitly in code of course, but it makes the code generator more complicated and the bytecode size significantly larger which affects early inlining behaviors (which are driven by bytecode size and not native code size).